Perform simple, systematic, stratified, cluster sampling.

Additional requirements:
(a) Ensure that we only have 10% of the records in each case. So, remove extra records, if necessary.
(b) For stratified, use strata as race.
(c) For cluster, use clusters as marital.status.

Then for overall data and for the four samples, find out the average education.num and average fnlwgt. Which of the four sampling methods is the closest to the overall average in the two cases?


In [2]:
import pandas as pd
import numpy as np

In [4]:
income_df = pd.read_csv('income.csv')
income_df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,40,Self-emp-not-inc,223881,Prof-school,15,Married-civ-spouse,Prof-specialty,Husband,White,Male,99999,0,70,United-States,>50K
1,30,Private,149118,HS-grad,9,Divorced,Craft-repair,Not-in-family,White,Female,0,0,40,United-States,<=50K
2,46,Private,109209,Some-college,10,Married-civ-spouse,Adm-clerical,Husband,White,Male,0,0,40,United-States,>50K
3,32,Private,229566,Assoc-voc,11,Married-civ-spouse,Other-service,Husband,White,Male,0,0,60,United-States,>50K
4,54,?,148657,Preschool,1,Married-civ-spouse,?,Wife,White,Female,0,0,40,Mexico,<=50K


In [5]:
orignal_row_count = len(income_df)
print(orignal_row_count)

25000


In [6]:
#simple random sampling
simple_df = income_df.sample(frac = 0.1, random_state = 101)
print(simple_df.head())
simple_row_count = len(simple_df)
print("Simple row count", simple_row_count)

       age         workclass  fnlwgt  education  education.num  \
7799    34  Self-emp-not-inc  288486       11th              7   
4427    32           Private   90969  Assoc-voc             11   
14941   31           Private  109055    HS-grad              9   
11644   40         State-gov   50093    Masters             14   
15548   48           Private  167159  Assoc-voc             11   

           marital.status        occupation relationship   race     sex  \
7799   Married-civ-spouse      Craft-repair      Husband  White    Male   
4427   Married-civ-spouse  Transport-moving      Husband  White    Male   
14941  Married-civ-spouse   Farming-fishing      Husband  White    Male   
11644            Divorced    Prof-specialty    Unmarried  White  Female   
15548       Never-married      Adm-clerical    Unmarried  White    Male   

       capital.gain  capital.loss  hours.per.week native.country income  
7799              0             0              40  United-States  <=50K  
4427

In [7]:
#systematic random sampling
system_df = income_df.iloc[::10, :]
print(system_df.head())
print("Systematic Row Count", len(system_df))

    age         workclass  fnlwgt     education  education.num  \
0    40  Self-emp-not-inc  223881   Prof-school             15   
10   21           Private   57916       HS-grad              9   
20   26           Private  208122  Some-college             10   
30   23           Private  139012     Assoc-voc             11   
40   19           Private   25429          12th              8   

        marital.status         occupation   relationship                race  \
0   Married-civ-spouse     Prof-specialty        Husband               White   
10       Never-married    Protective-serv      Own-child               White   
20            Divorced  Handlers-cleaners  Not-in-family               White   
30       Never-married   Transport-moving      Own-child  Asian-Pac-Islander   
40       Never-married      Other-service      Own-child               White   

       sex  capital.gain  capital.loss  hours.per.week native.country income  
0     Male         99999             0     

In [8]:
#stratified sampling on race
race = income_df['race'].unique()
print(race)

['White' 'Black' 'Asian-Pac-Islander' 'Other' 'Amer-Indian-Eskimo']


In [9]:
total_white = len(income_df[income_df['race'] == 'White'])
total_black = len(income_df[income_df['race'] == 'Black'])
total_Asian = len(income_df[income_df['race'] == 'Asian-Pac-Islander'])
total_Amer = len(income_df[income_df['race'] == 'Amer-Indian-Eskimo'])
total_others = len(income_df[income_df['race'] == 'Other'])

sample_white = int(0.1 * total_white)
sample_black = int(0.1 * total_black)
sample_asian = int(0.1 * total_Asian)
sample_amer = int(0.1 * total_Amer)
sample_other = int(0.1 * total_others)

print( "White : ", total_white, sample_white)
print("Black : ", total_black, sample_black)
print("Asian : ", total_Asian, sample_asian)
print("American : ", total_Amer, sample_amer)
print("Others : ", total_others, sample_other)

White :  21328 2132
Black :  2412 241
Asian :  799 79
American :  252 25
Others :  209 20


In [10]:
white_sample = income_df[income_df['race'] == 'White'].sample(n=sample_white, random_state = 101)
black_sample = income_df[income_df['race'] == 'Black'].sample(n=sample_black, random_state = 101)
asian_sample = income_df[income_df['race'] == 'Asian-Pac-Islander'].sample(n=sample_asian, random_state = 101)
amer_sample = income_df[income_df['race'] == 'Amer-Indian-Eskimo'].sample(n=sample_amer, random_state = 101)
other_sample = income_df[income_df['race'] == 'Other'].sample(n=sample_other, random_state = 101)

In [11]:
stratified_sample = pd.concat([white_sample, black_sample, asian_sample, amer_sample, other_sample])
stratified_len = len(stratified_sample)
print(stratified_len)

2497


In [1]:
#g_race = income_df.groupby('race')
#df_race = g_race.apply(lambda x: x.sample(frac=0.1,random_state=42))
#len(df_race)

In [12]:
#Cluster sampling on marital status
marital_status = income_df['marital.status'].unique()
marital_status

array(['Married-civ-spouse', 'Divorced', 'Widowed', 'Never-married',
       'Separated', 'Married-spouse-absent', 'Married-AF-spouse'],
      dtype=object)

In [13]:
np.random.seed(42)
selected_marital_status = np.random.choice(marital_status, size = 3, replace = False)
print(selected_marital_status)

['Married-civ-spouse' 'Divorced' 'Married-spouse-absent']


In [21]:
sample1 = income_df[income_df['marital.status'] == 'Married-civ-spouse'].sample(frac = 0.17, random_state = 101)
sample2 = income_df[income_df['marital.status'] == 'Divorced'].sample(frac = 0.17, random_state = 101)
sample3 = income_df[income_df['marital.status'] == 'Married-spouse-absent'].sample(frac = 0.17, random_state = 101)

#m = cluster_df[cluster_df['marital.status']== 'Married-civ-spouse']
#d = cluster_df[cluster_df['marital.status']== 'Divorced']
#m1 = cluster_df[cluster_df['marital.status']== 'Married-spouse-absent']
#print(len(m),len(d),len(m1))
cluster_df = pd.concat([sample1,sample2,sample3])
cluster_df.head(10)
len(cluster_df)

2587

In [15]:
#Overall average
avg_overall_df_edu = round(income_df['education.num'].mean(),2)
avg_overall_df_fnlwgt = round(income_df['fnlwgt'].mean(),2)
print(f"Avg of education.num = {avg_overall_df_edu:.2f}")
print(f"Avg of fnlwgt = {avg_overall_df_fnlwgt:.2f}")

Avg of education.num = 10.08
Avg of fnlwgt = 189661.13


In [16]:
#Simple sampling average
avg_simple_df_edu = round(simple_df['education.num'].mean(),2)
avg_simple_df_fnlwgt = round(simple_df['fnlwgt'].mean(),2)
print(f"Avg of education.num = {avg_simple_df_edu:.2f}")
print(f"Avg of fnlwgt = {avg_simple_df_fnlwgt:.2f}")

Avg of education.num = 10.12
Avg of fnlwgt = 191149.85


In [17]:
#Systematic sampling average
avg_system_df_edu = round(system_df['education.num'].mean(),2)
avg_system_df_fnlwgt = round(system_df['fnlwgt'].mean(),2)
print(f"Avg of education.num = {avg_system_df_edu :.2f}")
print(f"Avg of fnlwgt = {avg_system_df_fnlwgt:.2f}")

Avg of education.num = 10.05
Avg of fnlwgt = 188785.60


In [18]:
#Stratified sampling average
avg_strat_df_edu = round(stratified_sample['education.num'].mean(),2)
avg_strat_df_fnlwgt = round(stratified_sample['fnlwgt'].mean(),2)
print(f"Avg of education.num = {avg_strat_df_edu :.2f}")
print(f"Avg of fnlwgt = {avg_strat_df_fnlwgt:.2f}")

Avg of education.num = 10.04
Avg of fnlwgt = 187124.45


In [22]:
#Cluster sampling average
avg_cluster_df_edu = round(cluster_df['education.num'].mean(),2)
avg_cluster_df_fnlwgt = round(cluster_df['fnlwgt'].mean(),2)
print(f"Avg of education.num = {avg_cluster_df_edu:.2f}")
print(f"Avg of fnlwgt = {avg_cluster_df_fnlwgt:.2f}")

Avg of education.num = 10.24
Avg of fnlwgt = 188835.19


Overall average: 10.08 & 189661.13
Systematic average: 10.05 & 188785.60
Systematic Sampling's education.num and fnlwgt averages are closer to overall average as compared to other samplings.